docs: update SSO docs with latest CLI commands (#13272)
supabase/supabase: apps/docs/pages/guides/auth/sso/auth-sso-saml.mdx

export const meta = {
  id: 'auth-sso-saml',
  title: 'Single Sign-On with SAML 2.0 for Projects',
  description: 'Use Single Sign-On (SSO) authentication on your project with SAML 2.0',
  video: 'https://www.youtube.com/v/em1cpOAXknM',
}

Supabase Auth supports enterprise-level Single Sign-On (SSO) for any identity providers compatible with the using the SAML 2.0 protocol. This is a non-exclusive list of supported identity providers:

- Google Workspaces (formerly known as GSuite)
- Okta, Auth0
- Microsoft Active Directory, Azure Active Directory, Microsoft Entra
- PingIdentity
- OneLogin

If you're having issues with identity provider software not on this list, please [open a support ticket](https://app.supabase.com/support/new).

## Prerequisites

> This guide requires the use of the [Supabase CLI](/docs/guides/cli). Please make sure you're using version **TODO** or higher.

Author: Change the version here!

export const meta = {
  id: 'auth-sso-saml',
  title: 'Single Sign-On with SAML 2.0 for Projects',
  description: 'Use Single Sign-On (SSO) authentication on your project with SAML 2.0',
  video: 'https://www.youtube.com/v/em1cpOAXknM',
}

Supabase Auth supports enterprise-level Single Sign-On (SSO) for any identity providers compatible with the using the SAML 2.0 protocol. This is a non-exclusive list of supported identity providers:

- Google Workspaces (formerly known as GSuite)
- Okta, Auth0
- Microsoft Active Directory, Azure Active Directory, Microsoft Entra
- PingIdentity
- OneLogin

If you're having issues with identity provider software not on this list, please [open a support ticket](https://app.supabase.com/support/new).

## Prerequisites

> This guide requires the use of the [Supabase CLI](/docs/guides/cli). Please make sure you're using version v1.46.0 or higher. You can use `supabase -v` to see the currently installed version.

Reviewer: ```suggestion
This guide requires the use of the [Supabase CLI](/docs/guides/cli). Please make sure you're using version v1.46.4 or higher. You can use `supabase -v` to see the currently installed version.
```
Probably should update this once dogfooding is done, guessing that it might be v1.46.4 after any patches

Once you've obtained the SAML 2.0 Metadata XML file or URL you can establish a connection with your project's Supabase Auth server by using the [Supabase CLI](/docs/guides/cli):

```bash
supabase sso add --type saml --project-ref <your-project> \
  --metadata-url 'https://company.com/idp/saml/metadata' \
  --domains company.com
```

If you wish to use a Metadata XML file instead, you can use:

```bash
supabase sso add --type saml --project-ref <your-project> \
  --metadata-file /path/to/saml/metadata.xml \
  --domains company.com
```

This command will register a new identity provider with your project's Auth server. When successful, you will see the details of the provider such as it's SAML information and registered domains.

Please note that only persons with write access to the project can register, update or remove identity providers.

Once you've added an identity provider, users who have access to it can sign in to your application. With SAML 2.0 there are two ways that users can sign in to your project:

- By signing-in from your application's user interface, commonly known as **SP (Service Provider) Initiated Flow**

> - By clicking on an icon in the application menu on the company intranet or identity provider page, commonly known as **Identity Provider Initiated (IdP) Flow**

Reviewer 2: supported but seems quite tricky to set up - will probably need to branch off to a separate guide here in the future

- Department or organization
- Other attributes present in the users directory managed by the identity provider

> Other than the unique ID of the user, SAML does not make it mandatory that any other attributes appear in the assertion. Identity providers are configured about what user information is shared with your project.

Reviewer: /nit 
```suggestion
With exception of the unique user ID, SAML does not require any other attributes in the assertion. Identity providers can be configured so that only select user information is shared with your project.
```

}
```

When creating or updating an identity provider with the [Supabase CLI](/docs/guides/cli) you can include this JSON as a file with the `--attribute-mapping /path/to/attribute/mapping.json` flag.

For example, to change the attribute mappings to an existing provider you can use:

```bash

> supabase sso update <provider-uuid> --project-ref <your-project> \

Reviewer: On v1.46.1 the flag is labelled as  `--attribute-mapping-file` instead of `--attribute-mapping` do double check before merging

Reviewer: ![CleanShot 2023-03-30 at 20 12 18@2x](https://user-images.githubusercontent.com/8011761/228832094-9aa7f57b-c8d7-45e0-9efa-d53aacabc755.png) for quick reference

Reviewer 2: should be `--attribute-mapping-file` based on the latest version

At this time it is not possible to have users without an email address, so SAML assertions without one will be rejected.

Most SAML 2.0 identity providers use LDAP attribute names. However, due to their variability and complexity operators of identity providers are able to customize both the `Name` and attribute value that is sent to Supabase Auth in an assertion. Please refer to the identity provider's documentation and contact the operator for details on what attributes are mapped for your project.

### Remove a connection

Once a connection to an identity provider is established, you can remove it like so:

```bash
supabase sso remove <provider-id> --project-ref <your-project>
```

> When successful, you will see the details of the removed identity provider. All user accounts from that identity provider will be immediately logged out. User information will remain in the system, but it will no longer be possible for any of those accounts to be accessed in the future, even if you add the connection again.

Author: Let's leave it to see if people actually run into the problem and are not able to figure it out on their own. I doubt it's a common occurrence...

```bash
supabase sso show <provider-id> --project-ref <your-project>
```

You can use the `-o json` flag to output the information as JSON, should you need to. Other formats may be supported, please use `--help` to see all available options.

## Frequently Asked Questions

### How do I publish my application to an identity provider's marketplace?

Many cloud-based identity providers offer a marketplace where you can register your application for easy on-boarding with customers. When you use Supabase Auth's SAML 2.0 support you can register your project in any one of these marketplaces.

Please refer to the relevant documentation for each cloud-based identity provider on how you can do this. Some common marketplaces are:

- [Okta Integration Network](https://developer.okta.com/docs/guides/build-sso-integration/saml2/main/)
- [Azure Active Directory App Gallery](https://learn.microsoft.com/en-us/azure/active-directory-b2c/publish-app-to-azure-ad-app-gallery)
- [Google Workspaces Pre-integrated SAML apps catalog](https://support.google.com/a/table/9217027)

### How do I use SAML in local development?

> Unfortunately using SAML in when developing on your machine is pretty difficult. This is a security limitation by identity providers which do not allow setting the SAML Assertion Consumer Service URL to non-HTTPS URLs, or they completely disallow the use of `localhost` in such settings.

Reviewer: /nit consider rephrasing

```suggestion
Unfortunately, it is difficult to use SAML when developing on your machine. This is a security limitation by identity providers which do not allow setting the SAML Assertion Consumer Service URL to non-HTTPS URLs, or they completely disallow the use of `localhost` in such settings.
```

> You could get around these limitations by exposing your local Supabase stack to the internet and placing it behind a HTTPS server. Alternatively, you could use a locally hosted SAML 2.0 identity provider that will allow you to use insecure URLs.

Reviewer 2: would they be able to use `https://samltest.id/` for this if they want to try out the entire flow though?

Author: Not really, it's somewhat doable but it's a bit too complex to go into. We can maybe have a separate blog post about it -- testing SAML locally or something. I wouldn't go into it in this doc as it's very little value.

Author: This is why I wanted to create the saml tool, but haven't had enough time to finish it.

Reviewer 2: should we just remove this section then? not too sure if it provides much value if there aren't clear instructions on how to go about doing so
